src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java

```diff
@@ -0,0 +1,101 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api;
+
+import com.cloudbees.jenkins.plugins.bitbucket.endpoints.BitbucketCloudEndpoint;
+import com.cloudbees.plugins.credentials.common.StandardCredentials;
+import jenkins.authentication.tokens.api.AuthenticationTokenContext;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.HttpClientBuilder;
+
+/**
+ * Support for various different methods of authenticating with Bitbucket
+ */
+public abstract class BitbucketAuthenticator {
+
+    private String id;
+
+    /**
+     * The URL protocol as reported in an {@link AuthenticationTokenContext}
+     */
+    public static final String PROTOCOL_PURPOSE = "PROTOCOL";
```

Comment:
I would call this `Scheme`

Reply:
For the constant name, or its string value? Also, `<blah>_PURPOSE` reads kind of weird. Is there a naming scheme that would be more preferred than that?

Reply:
I’d use `SCHEME = “scheme”` or `SERVER_URI_SCHEME = “bitbucket.server.uri.scheme”`

The length of the constant doesn’t use up more space as it will still be an 8 byte reference to the constant pool anyway, so more descriptive is better